impeller/typographer/backends/stb/text_render_context_stb.cc

```diff
@@ -0,0 +1,587 @@
+#ifndef MAX_GLYPH_ATLAS_SIZE
+#define MAX_GLYPH_ATLAS_SIZE 4096u
+#endif
+#ifndef MIN_GLYPH_ATLAS_SIZE
+#define MIN_GLYPH_ATLAS_SIZE 8u
+#endif
+
+#define DISABLE_COLOR_FONT_SUPPORT 1
+#ifdef DISABLE_COLOR_FONT_SUPPORT
+#define COLOR_FONT_BPP 1
+#else
+#define COLOR_FONT_BPP 4
+#endif
+
+namespace impeller {
+
+// An simple bitmap in lieu of a skia bitmap.
+class STBBitmap {
+ public:
+  ~STBBitmap() = default;
+
+  STBBitmap() = delete;
+
+  STBBitmap(size_t width, size_t height, size_t bytes_per_pixel)
+      : width_(width),
+        height_(height),
+        bytes_per_pixel_(bytes_per_pixel),
+        pixels_(std::make_unique<uint8_t[]>(width * height * bytes_per_pixel)) {
+  }
+
+  uint8_t* GetPixels() const { return pixels_.get(); }
+
+  uint8_t* GetPixelAddress(TPoint<size_t> coords) const {
+    FML_DCHECK(coords.x < width_);
+    FML_DCHECK(coords.x < height_);
+
+    return &pixels_.get()[(coords.x + width_ * coords.y) * bytes_per_pixel_];
+  }
+
+  size_t GetRowBytes() const { return width_ * bytes_per_pixel_; }
+
+  size_t GetWidth() const { return width_; }
+
+  size_t GetHeight() const { return height_; }
+
+  size_t GetSize() const { return width_ * height_ * bytes_per_pixel_; }
+
+ private:
+  size_t width_;
+  size_t height_;
+  size_t bytes_per_pixel_;
+  std::unique_ptr<uint8_t[]> pixels_;
+};
+
+// Analogous to the bitmaps (one for each type) stored in each Atlas context.
+static auto alpha_bitmap =
+    std::make_shared<STBBitmap>(MIN_GLYPH_ATLAS_SIZE, MIN_GLYPH_ATLAS_SIZE, 1);
+static auto color_bitmap = std::make_shared<STBBitmap>(MIN_GLYPH_ATLAS_SIZE,
+                                                       MIN_GLYPH_ATLAS_SIZE,
+                                                       COLOR_FONT_BPP);
+
+static std::shared_ptr<STBBitmap> get_atlas_bitmap(
+    impeller::GlyphAtlas::Type type) {
+  switch (type) {
+    case impeller::GlyphAtlas::Type::kAlphaBitmap: {
+      return alpha_bitmap;
+      break;
+    }
+    case impeller::GlyphAtlas::Type::kColorBitmap: {
+      return color_bitmap;
+      break;
+    }
+  }
+}
+
+static void update_atlas_bitmap(std::shared_ptr<STBBitmap>& bitmap,
+                                impeller::GlyphAtlas::Type type) {
+  switch (type) {
+    case impeller::GlyphAtlas::Type::kAlphaBitmap: {
+      alpha_bitmap = bitmap;
+      break;
+    }
+    case impeller::GlyphAtlas::Type::kColorBitmap: {
+      color_bitmap = bitmap;
+      break;
+    }
+  }
+}
+
+using FontGlyphPairRefVector =
+    std::vector<std::reference_wrapper<const FontGlyphPair>>;
+
+constexpr size_t kPadding = 1;
+
+std::unique_ptr<TextRenderContext> TextRenderContextSTB::Make() {
+  return std::make_unique<TextRenderContextSTB>();
+}
+
+TextRenderContextSTB::TextRenderContextSTB() : TextRenderContext() {}
+
+TextRenderContextSTB::~TextRenderContextSTB() = default;
+
+// Function returns the count of "remaining pairs" not packed into rect of given
+// size.
+static size_t PairsFitInAtlasOfSize(
```

Comment:
We really need a generic rect packer ready to go in `//impeller/core`. Can we file an issue for this?

Reply:
https://github.com/flutter/flutter/issues/133028